### PR TITLE
Import of Selkies JS Interposer for WebRTC to Linux Joystick support

### LINKED ISSUE #55: Find a way to use Gamepad controllers without external container privileges
Currently, selkies-gstreamer does support gamepads only when the uinput device is provisioned. This is very cumbersome and risky in Kubernetes.

Instead of using that, it may be possible to directly translate them to Xorg operations (and later Wayland #46 using [virtual keyboard](https://wayland.app/protocols/virtual-keyboard-unstable-v1)).

Relevant files: [gst-web](https://github.com/selkies-project/selkies-gstreamer/tree/master/addons/gst-web/src) module and [webrtc_input.py](https://github.com/selkies-project/selkies-gstreamer/blob/master/src/selkies_gstreamer/webrtc_input.py).

### LINKED ISSUE #90: Add instructions to use uinput for bare metal/Docker instances, not just GKE or Kubernetes
The uinput interface is currently undocumented for anything other than uinput-device-plugin (for Kubernetes). Docs/adequate code should be added for normal OS or Docker interfaces.

### PR DESCRIPTION
**Pull request by danisla**
Fixes #90
Fixes #55

## New features
Import of Selkies JS Interposer for WebRTC to Linux Joystick support without uinput dependency
- LD_PRELOAD syscall interposer to emulate gamepads.
- Intercepts open() requests to /dev/input/jsX devices.
- Interposer connects to selkies unix domain socket.
- Interposer writes log to /tmp/selkies_js.log
- Selkies WebRTC app sends Javascript gamepad events to socket.
- Up to 4 simultaneous gamepads supported from WebRTC.
- Build as debian package and installed to /usr/local/lib/selkies-js-interposer/joystick_interposer.so
- Dev container updated to auto-load interposer.
- Update to github workflow to build and publish image and release asset.
- Add joystick interposer to example Dockerfile

Add turnserver to example docker image
- Helps with initial connectivity in environments like WSL with Docker.
- Can be overridden with env vars or cli args.

## Supported Gamepads
I've tested this with apps like retroarch and it seems to work well.

Currently only Xpad like gamepads are emulated for the button and axes mapping. 

I've tested with the following controllers:
- Xbox Series X/S Controller
- Stadia Controller
- 8Bitdo SN30 PRO

Others will surely work just fine, especially if they have similar button layouts to an Xpad.

Please let me know if there is a gamepad that doesn't work and we can try to accommodate.

## Testing
The `testing` branch has been built with the new joystick interposer pre-installed and added to the default LD_PRELOAD environment variable. This branch also contains the integrated turnserver to the example container, hence the new port 3478 to the docker run command.

You can test it with the example docker container like this:

Ubuntu 22.04
```
docker run --pull=always --name selkies -it --rm -p 8080:8080 -p 3478:3478 ghcr.io/selkies-project/selkies-gstreamer/gst-py-example:testing-ubuntu22.04
```

For Ubuntu 20.04:
```
docker run --pull=always --name selkies -it --rm -p 8080:8080 -p 3478:3478 ghcr.io/selkies-project/selkies-gstreamer/gst-py-example:testing-ubuntu20.04
```

## Usage
1. Connect to the web interface.
2. Connect a gamepad to your local computer.
3. Press any button on the gamepad to connect it to the web interface.
4. In the remote session, run a program like `jstest-gtk` to verify you see the `Selkies Controller`
5. Connect up to 4 total gamepads using the same method.
Selkies Gamepad Test

The example docker image has these environment variables set:
```bash
export LD_PRELOAD=/usr/local/lib/selkies-js-interposer/joystick_interposer.so
export SDL_JOYSTICK_DEVICE=/dev/input/js0
```
The entrypoint script also creates fake jsX files in /dev/input like below to help programs that try to list the /dev/input directory discover the joystick devices:
```
sudo mkdir -p /dev/input
sudo touch /dev/input/{js0,js1,js2,js3}
```

Some software programs use different methods to discover linux joysticks. The LD_PRELOAD method seems to work for most direct linux applications and the `SDL_JOYSTICK_DEVICE` env var helps with SDL applications.

### CODE REVIEW

**Comment by ehfd** [comment]
Wow. This actually works?...

One aspect of testing that should be done is with Wine games and applications.

**Comment by ehfd** [comment]
https://github.com/selkies-project/selkies-gstreamer/issues/28#issuecomment-1277196623

Could this approach be expanded to allow true relative mouse cursors?

**Comment by danisla** [comment]
> Wow. This actually works?...
> 
> One aspect of testing that should be done is with Wine games and applications.

If someone can help test, that would be great.

**Comment by danisla** [comment]
> [#28 (comment)](https://github.com/selkies-project/selkies-gstreamer/issues/28#issuecomment-1277196623)
> 
> Could this approach be expanded to allow true relative mouse cursors?

I think the approach would be similar, yes. My plan is to investigate doing this for mouse emulation later so we can completely remove the dependency on uinput and have full support in container environments.
However, Xorg has a funny way of loading core input devices like Mice and Keyboards so it might have to be a custom X11 module or a full uinput interposer. Either way, that's a separate task from the gamepad support in this PR.

**Comment by ehfd** [comment]
Good to know that the approach is similar with mouse devices.

Unfortunately, I don't have a real gamepad device at the moment and I am having a hard time with time management.

@xhejtman perhaps interested?
Or any of the community.

File: src/selkies_gstreamer/__main__.py [hunk 1]
```diff
@@ -383,9 +383,18 @@ def main():
     parser.add_argument('--uinput_mouse_socket',
                         default=os.environ.get('UINPUT_MOUSE_SOCKET', ''),
                         help='Path to the uinput mouse socket provided by the uinput-device-plugin, if not provided uinput is used directly.')
-    parser.add_argument('--uinput_js_socket',
-                        default=os.environ.get('UINPUT_JS_SOCKET', ''),
-                        help='Path to the uinput joystick socket provided by the uinput-device-plugin, if not provided uinput is used directly.')
+    parser.add_argument('--js0_socket',
```
Comment (ehfd) [inline src/selkies_gstreamer/__main__.py]: I think there is something better than to add four command line arguments. Would prefer that one command line argument that parses a maximum of four sockets.

File: src/selkies_gstreamer/webrtc_input.py [hunk 2]
```diff
@@ -650,7 +677,8 @@ def on_message(self, msg):
             self.on_set_enable_audio(enabled)
         elif toks[0] == "_arg_resize":
             if len(toks) != 3:
-                logger.error("invalid _arg_resize commnad, expected 2 arguments <enabled>,<resolution>")
+                logger.error(
+                    "invalid _arg_resize commnad, expected 2 arguments <enabled>,<resolution>")
```
Comment (ehfd) [inline src/selkies_gstreamer/webrtc_input.py]: Typo.

**Review by ehfd — COMMENTED** [review]
Overall, looks okay but was just skimming. Made comments.

**Comment by ehfd** [comment]
This commit is a great milestone which is a step forward from all known Linux remote desktop implementations working in containers.

**Pull request merged**